frame/evm/src/lib.rs

```diff
@@ -508,6 +513,17 @@ pub mod pallet {
 	#[pallet::getter(fn account_storages)]
 	pub type AccountStorages<T: Config> =
 		StorageDoubleMap<_, Blake2_128Concat, H160, Blake2_128Concat, H256, H256, ValueQuery>;
+
+	/// Allows for precompiles to have arbitrary addresses, potentially more than one.
+	/// `k1`: precompile label, e.g.: `b"Sha3FIPS256".to_vec()`
+	/// `k2`: precompile address
+	///
+	/// Please note that adding a new precompile label here is not enough to guarantee its execution
+	/// It is also required to list the new label on the implementation of `PrecompileSet::execute()`
+	#[pallet::storage]
+	#[pallet::getter(fn precompiles)]
+	pub type Precompiles<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, Vec<u8>, Blake2_128Concat, H160, (), OptionQuery>;
```

Comment:
If this pr is finally acceptable, consider changing the `()` to `bool`? 
This way, you can choose to enable/disable one of the precompile's addresses.

Reply:
@sorpaas I switched to `StorageMap`.

But then I was thinking: wouldn't it be useful to use `StorageDoubleMap` and have an index for each `H160` address?
- `k1`: `H160`
- `k2`: `PrecompileLabel`
- `v`: `PrecompileIndex`

where `PrecompileIndex` is a `Config` type, with same traits as [`AssetId`](https://github.com/paritytech/substrate/blob/33eee69bf53d45449c1f6b145afe5da872310b6a/frame/assets/src/lib.rs#L239)

that would avoid the need for using [`#[precompile_set]`](https://github.com/PureStake/moonbeam/blob/master/precompiles/utils/macro/docs/precompile_macro.md#precompileset) to distinguish between precompiles within the same label.

Reply:
cool, I'll refactor the changes into a separate `pallet-evm-precompile`

Reply:
do we leave `template` using the old `FrontierPrecompiles` as `PrecompileSet`, or the new one (`pallet-evm-precompile`)?

Reply:
I did some experiments and it turns out having an on-chain `PrecompileIndex` is not really useful.

```
impl<R> PrecompileSet for OnChainPrecompiles<R>
where
	R: pallet::Config + pallet_evm::Config,
{
	fn execute(&self, handle: &mut impl PrecompileHandle) -> Option<PrecompileResult> {
		match handle.code_address() {
			// Ethereum precompiles :
			a if &Precompiles::<R>::get(a).0[..] == b"PrecompileX" => {
				Some(PrecompileX::execute(handle))
			}
...
```
even though I can do something like `let precompile_index = Precompiles::<R>::get(a).1`, I have no means to pass it to `PrecompileX::execute(handle)`.

It seems the only available approach is using Moonbeam's `#[precompiles::precompile_set] & #[precompile::discriminant]`.